pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,576 @@
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the exception in its prelude.
+
+```ebnf {% showTitle=false %}
+Exception
+    : Prelude "exception" Identifier (":" TypeRef)? "{" UndelimitedList<Field> "}"
+    ;
+```
+
+For additional information on exceptions, see the [exception][exception-guide] page.
+
+### Fields
+
+Fields can only be declared within [structs][struct], [classes][class], and [exceptions][exception].
+Field declarations consist of an [identifier], followed by a colon, and then a [type].
+Optionally, a [tag] may be applied to the field, directly before its identifier, making this a [tagged field][tag].
+Additionally, [local attributes][attribute] and a [doc-comment][doc-comment] may be applied to the field in its prelude.
+
+```ebnf {% showTitle=false %}
+Field
+    : Prelude Tag? Identifier ":" TypeRef
+    ;
+```
+
+For additional information on fields, see the [field][field-guide] page.
+
+### Interfaces
+
+Interface definitions consist of the `interface` keyword, followed by an [identifier], optionally followed by a list of base interfaces, and then the interface's body.
+Base interfaces are specified by a single colon, followed by the (possibly scoped) identifiers of one or more interfaces, separated by commas.
+Interface bodies consist of a set of [operations][operation] wrapped in a pair of braces.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the class in its prelude.
+
+```ebnf {% showTitle=false %}
+Interface
+    : Prelude "interface" Identifier (":" NonEmptyCommaList<TypeRef>)? "{" Operation* "}"
+    ;
+```
+
+If you are using an RPC framework, every interface definition also defines a _proxy type_.
+This proxy type shares the same name and scope as its corresponding interface and can be used as a type.
+
+```slice
+interface Foo { // Defines an interface named `Foo`.
+    op() -> Foo // Returns a proxy named `Foo`.
+}
+```
+
+It is illegal to use an interface as a type without an RPC framework.
+
+For additional information on interfaces and proxies, see the [interface][interface-guide] and [proxy][proxy-guide] pages.
+
+### Operations
+
+An operation consists of an [identifier], followed be a list of [parameters][parameter], optionally followed by a return type, optionally followed by an [exception specification][exception-specification-guide].
+Operations support the `idempotent` modifier keyword. For information about its effects, see the [idempotent][idempotent-guide] section.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the operation in its prelude.
+
+```ebnf {% showTitle=false %}
+Operation
+    : Prelude "idempotent"? Identifier "(" UndelimitedList<Parameter> ")" ReturnType? ExceptionSpecification?
+    ;
+```
+
+There are two syntaxes for specifying a return type, depending on the number of elements you're returning.
```

Comment:
I think return types should be documented as part of Parameters, not within Operation(s).

The syntax for return parameters is largely the same as the syntax for operation parameters, and the Language Guide describe them together:
https://docs.icerpc.dev/slice2/language-guide/parameters

Reply:
I tried to do this, but it's awkward.
Yes, they're very similar to parameters, but the grammar rules are totally separate.

The 'special' parameter syntax is baked into the rule for `ReturnType`s:
```ebnf {% showTitle=false %}
ReturnType
    : "->" Tag? "stream"? TypeRef
    | "->" "(" UndelimitedList<Parameter> ")"
    ;
```

And I think it makes the most sense to document the concept of "return types" on operations, instead of parameters.

Note that we have a sentence and link mentioning this in the documentation though:
`The syntax of a single return type is identical to a [parameter], but without an identifier or colon separator.`

Reply:
> And I think it makes the most sense to document the concept of "return types" on operations, instead of parameters.

So the documentation for operations includes the description for return types and references the documentation for parameters (which are at the same level as operations) even though both parameters and return types are used only in operations.

This definitely makes complete sense!

Reply:
> even though both parameters and return types are used only in operations.

Yes, and enumerators are only used in enums, what's your point?
The documentation follows the grammar. As much as you seem to dislike it, the grammar of single return types and parameters _is different_.

> which are at the same level as operations

You really need to make up your mind about whether I should continue with this PR as-is or restructure it.